src/ceph-volume/ceph_volume/util/metadata.py

```diff
@@ -0,0 +1,438 @@
+    # Post process the LVM output
+    for lvm_info in lvm_data['report'][0][query_type]:
+        if query_type == 'lv':
+            key = lvm_info.get('lv_dm_path').replace('/dev/mapper/', '')
+        elif query_type == 'pv':
+            key = lvm_info.get('pv_name')
+        elif query_type == 'vg':
+            key = lvm_info.get('vg_name')
+        else:
+            raise ValueError("Call to get_lvm_data is invalid, must use lv, pv or vg query_type")
+
+        lvm[key] = lvm_info
+    return lvm
+
+
+def worker():
+    """worker function that takes a task from a queue, executes it and updates the shared data object
+    """
+    active = True
+    while active:
+        try:
+            task = task_queue.get_nowait()
+        except queue.Empty:
+            time.sleep(0.005)
+        else:
+            if not task:
+                active = False
+            else:
+                var_name, func, args = task
+                res = func(*args)
+                if '.' in var_name:
+                    assert var_name.count('.') == 1, "variable name passed with > 1 '.' separator"
+                    attr, key = var_name.split('.')
+                    with shared_data_lock:
+                        d = getattr(shared_data, attr)
+                        d.update({key: res})
+                else:
+                    with shared_data_lock:
+                        setattr(shared_data, var_name, res)
+                task_queue.task_done()
+
+
+def get_holders(dev_name: str) -> List[str]:
+    """Use the holder directory is sysfs to identify child devices
+
+    Args:
+        dev_name (str): device id e.g. sdb
+
+    Returns:
+        List[str]: List of devices related to the given device
+    """
+    holders = set()
+    dev_root = f"/sys/block/{dev_name}"
+    holder_dir_list = [f"{dev_root}/holders"]
+    holder_dir_list.extend(glob.glob(f"{dev_root}/**/holders"))
+    for holder_dir in holder_dir_list:
+        for dm_dev_id in os.listdir(holder_dir):
+            dm_name = get_file_contents(f"/sys/block/{dm_dev_id}/dm/name")
+            holders.add(dm_name)
+
+    return list(holders)
+
+
+def _js_reformat(raw_data: List[str]) -> Dict[str, Any]:
+    """reformat command output into json
+
+    Args:
+        raw_data (List[str]): Each string is a line of output
+
+    Returns:
+        Dict[str, Any]: JSON representation of the output
+    """
+    js_items = []
+    for j in raw_data:
+        js_items.append(j.strip())
+
+    try:
+        js = json.loads(''.join(js_items))
+    except json.JSONDecodeError:
+        # can't continue
+        raise
+
+    return js
+
+
+class MetadataCache:
+    """Create a cache of device metadata
+
+    Uses worker threads as a producer/consumer pool where tasks to run are placed
+    on a global queue. The worker threads can then process the task, and update the
+    shared_data object
+
+    Raises:
+        ValueError: [description]
+
+    Returns:
+        [type]: [description]
+    """
```

Comment:
This is great info - thanks! As far as the goals of this PR is concerned, perhaps cache is a misnomer.The goal is to create a metadata 'map' in one hit, whereas at the moment each device incurs multiple LVM, udevadm,blkid,lsblk calls - which is where the latency is coming from. 

So yes I did profile where inventory was spending it's time, and also found that not only are these calls serial, some of them return nothing anyway, since the parameters passed are invalid (for example lvs -S lv_path=/dev/sdb is not a valid lv pathname, and yet the code calls it erroneously.